src/sage/rings/valuation/developing_valuation.py

```diff
@@ -0,0 +1,310 @@
+    Abstract base class for a discrete valuation of polynomials defined over
+    the polynomial ring ``domain`` by the `\phi`-adic development.
+
+    EXAMPLES::
+
+        sage: R.<x> = QQ[]
+        sage: v = valuations.GaussValuation(R, QQ.valuation(7))
+
+    TESTS::
+
+        sage: TestSuite(v).run() # long time
+
+    """
+    def __init__(self, parent, phi):
+        r"""
+        TESTS::
+
+            sage: R.<x> = QQ[]
+            sage: v = valuations.GaussValuation(R, QQ.valuation(7))
+            sage: from sage.rings.valuation.developing_valuation import DevelopingValuation
+            sage: isinstance(v, DevelopingValuation)
+            True
+
+        """
+        DiscretePseudoValuation.__init__(self, parent)
+
+        domain = parent.domain()
+        from sage.rings.polynomial.polynomial_ring import is_PolynomialRing
+        if not is_PolynomialRing(domain) or not domain.ngens() == 1:
+            raise TypeError("domain must be a univariate polynomial ring but %r is not"%(domain,))
+
+        phi = domain.coerce(phi)
+        if phi.is_constant() or not phi.is_monic():
+            raise ValueError("phi must be a monic non-constant polynomial but %r is not"%(phi,))
+
+        self._phi = phi
+
+    def phi(self):
+        r"""
+        Return the polynomial `\phi`, the key polynomial of this valuation.
+
+        EXAMPLES::
+
+            sage: R = Zp(2,5)
+            sage: S.<x> = R[]
+            sage: v = GaussValuation(S)
+            sage: v.phi()
+            (1 + O(2^5))*x
+
+        """
+        return self._phi
+
+    def effective_degree(self, f, valuations=None):
+        r"""
+        Return the effective degree of ``f`` with respect to this valuation.
+
+        The effective degree of `f` is the largest `i` such that the valuation
+        of `f` and the valuation of `f_i\phi^i` in the development `f=\sum_j
+        f_j\phi^j` coincide (see [ML1936'] p.497.)
+
+        INPUT:
+
+        - ``f`` -- a non-zero polynomial in the domain of this valuation
+
+        EXAMPLES::
+
+            sage: R = Zp(2,5)
+            sage: S.<x> = R[]
+            sage: v = GaussValuation(S)
+            sage: v.effective_degree(x)
+            1
+            sage: v.effective_degree(2*x + 1)
+            0
+
+        """
+        f = self.domain().coerce(f)
+
+        if f.is_zero():
+            raise ValueError("the effective degree is only defined for non-zero polynomials")
+
+        if valuations is None:
+            valuations = list(self.valuations(f))
+        v = min(valuations)
+        return [i for i,w in enumerate(valuations) if w == v][-1]
+
+    @cached_method
+    def _pow(self, x, e, error, effective_degree):
+        r"""
+        Return `x^e`.
+
+        This method does not compute the exact value of `x^e` but only an
+        element that differs from the correct result by an error with valuation
+        at least ``error``. The output is assumed to have at most
+        ``effective_degree``.
```

Comment:
Clarify that output wouldn't be correct if the effective_degree given is incorrect (too low).